server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -4368,4 +4386,16 @@ public void waitForPrimaryTermAndGeneration(long primaryTerm, long segmentGenera
         );
     }
 
+    /**
+     * Synchronously closes the provided engine
+     */
+    private static void closeEngine(Engine engine) throws IOException {
+        if (engine != null) {
+            // TODO Ideally we would convert `closeEngine` calls to async `close` calls and remove UnsafePlainActionFuture
+            final var future = new UnsafePlainActionFuture<Void>(ThreadPool.Names.GENERIC);
```

Comment:
Introducing a new `UnsafePlainActionFuture` here doesn't really improve things much. Could we instead first make all the engine-closing calls take a listener (but in practice remain synchronous) with one or more preliminary PRs, and then we can flip it to an async implementation without fuss.

Reply:
I hated to use it here, too. Indeed, let me try to get rid of all synchronous calls in a separate PR :+1: